google/cloud/bigtable/_row_merger.py

```diff
@@ -0,0 +1,621 @@
+            }
+        else:
+            # remove seen keys from user-specific key list
+            row_keys: list[bytes] = row_set.get("row_keys", [])
+            adjusted_keys = []
+            for key in row_keys:
+                if key not in emitted_rows:
+                    adjusted_keys.append(key)
+            # if user specified only a single range, set start to the last seen key
+            row_ranges: list[dict[str, Any]] = row_set.get("row_ranges", [])
+            if len(row_keys) == 0 and len(row_ranges) == 1:
+                row_ranges[0]["start_key_open"] = last_seen_row_key
+                if "start_key_closed" in row_ranges[0]:
+                    row_ranges[0].pop("start_key_closed")
+            return {"row_keys": adjusted_keys, "row_ranges": row_ranges}
+
+    @staticmethod
+    async def merge_row_response_stream(
+        request_generator: AsyncIterable[ReadRowsResponse], state_machine: _StateMachine
+    ) -> AsyncGenerator[Row | RequestStats, None]:
+        """
+        Consume chunks from a ReadRowsResponse stream into a set of Rows
+
+        Args:
+          - request_generator: AsyncIterable of ReadRowsResponse objects. Typically
+                this is a stream of chunks from the Bigtable API
+        Returns:
+            - AsyncGenerator of Rows
+        Raises:
+            - InvalidChunk: if the chunk stream is invalid
+        """
+        async for row_response in request_generator:
+            # unwrap protoplus object for increased performance
+            response_pb = row_response._pb
+            last_scanned = response_pb.last_scanned_row_key
+            # if the server sends a scan heartbeat, notify the state machine.
+            if last_scanned:
+                yield state_machine.handle_last_scanned_row(last_scanned)
+            # process new chunks through the state machine.
+            for chunk in response_pb.chunks:
+                complete_row = state_machine.handle_chunk(chunk)
+                if complete_row is not None:
+                    yield complete_row
+            # yield request stats if present
+            if row_response.request_stats:
+                yield row_response.request_stats
+        if not state_machine.is_terminal_state():
+            # read rows is complete, but there's still data in the merger
+            raise InvalidChunk("read_rows completed with partial state remaining")
+
+
+class _StateMachine:
+    """
+    State Machine converts chunks into Rows
+
+    Chunks are added to the state machine via handle_chunk, which
+    transitions the state machine through the various states.
+
+    When a row is complete, it will be returned from handle_chunk,
+    and the state machine will reset to AWAITING_NEW_ROW
+
+    If an unexpected chunk is received for the current state,
+    the state machine will raise an InvalidChunk exception
+    """
+
+    def __init__(self):
+        self.completed_row_keys: Set[bytes] = set({})
```

Comment:
Are we keeping all the completed_row_keys in memory? This would end in high memory usage.I think check the order of the row_keys is enough.

Reply:
I'll take a look at this!